.travis.yml

```diff
@@ -16,19 +17,45 @@ matrix:
 
 before_install:
   - mkdir -p $HOME/.local/bin
-  - export PATH=$HOME/.local/bin:$PATH
+  - export PATH=/opt/ghc/$GHCVER/bin:/opt/cabal/$CABALVER/bin:$HOME/.local/bin:$PATH
   - curl -L https://github.com/commercialhaskell/stack/releases/download/v1.0.4.3/stack-1.0.4.3-linux-x86_64.tar.gz | tar xz --wildcards --strip-components=1 -C $HOME/.local/bin '*/stack'
   - curl -L https://github.com/google/protobuf/releases/download/v3.0.0-beta-2/protoc-3.0.0-beta-2-linux-x86_64.zip > protoc-release.zip
   - unzip -p protoc-release.zip protoc > $HOME/.local/bin/protoc
   - chmod a+x $HOME/.local/bin/protoc
   - rm protoc-release.zip
 
 install:
-  - stack setup --no-terminal
-  - stack build --only-snapshot --no-terminal
+  # - stack setup --no-terminal
+  # - stack build --only-snapshot --no-terminal
+  - cabal --version
+  - cabal update
 
 script:
   # Separate build from test, since build by itself hits some edge cases around
   # custom Setup.hs script dependencies.
-  - stack build --haddock --no-haddock-deps
-  - stack test --haddock --no-haddock-deps
+  # - stack build --haddock --no-haddock-deps
+  # - stack test --haddock --no-haddock-deps
+  - PACKAGES=$(stack query | sed -n 's/  \(.*\):$/\1/p')
+  - for p in $PACKAGES;
+    do
+      cd $p;
+      cabal install --only-dependencies;
+      cabal install happy;
+      cabal configure --enable-tests;
+      cabal build;
+      cabal sdist;
+      cd ..;
+    done
+  - for p in $PACKAGES;
+    do
+      cd $p;
+      SRC_TGZ=$(cabal info . | awk '{print $2 ".tar.gz";exit}');
+      cd dist;
+      if [ -f "$SRC_TGZ" ]; then
+        cabal install --force-reinstalls "$SRC_TGZ";
+        else
+          echo "expected '$SRC_TGZ' not found";
+          exit 1;
+      fi;
+      cd ..;
```

Comment:
This can be avoided by nesting shells:
for p in $PACKAGES;
 +    do
 +      ( cd $p && 
 +      ...
         )